internal/database/autocontact.go

```diff
@@ -0,0 +1,169 @@
+package database
+
+import (
+	"log/slog"
+	"net"
+	"os"
+	"time"
+
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/net/icmp"
+	"golang.org/x/net/ipv4"
+)
+
+type SSHConfig struct {
+	User    string
+	Keypath string
+}
+
+func MonitorForDeadMachines(interval int, database Database, timespanForDeath int, l *slog.Logger, s SSHConfig) error {
+	downsampleTicker := time.NewTicker(time.Duration(interval))
+
+	for t := range downsampleTicker.C {
+		err := monitor(database, t, timespanForDeath, l, s)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func monitor(database Database, t time.Time, timespanForDeath int, l *slog.Logger, s SSHConfig) error {
+	last_seens, err := database.LastSeen()
+
+	if err != nil {
+		return err
+	}
+
+	for idx := range last_seens {
+		seen := last_seens[idx]
+
+		if seen.LastSeen < t.Add(-1*time.Duration(timespanForDeath)*time.Second).Unix() {
+			if ping(seen.Hostname, l) {
+				err := sshRestart(seen.Hostname, l, s)
+
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func sshRestart(remote string, l *slog.Logger, s SSHConfig) error {
+	user := s.User
+	keypath := s.Keypath
+
+	key, err := os.ReadFile(keypath)
+	if err != nil {
+		l.Error("Unable to read key: %v", err)
```

Comment:
NIT: `log/slog` implements an idea called structured logging, where all attributes are key-value pairs. The correct way to do this is:

```go
l.Error("Unable to read ssh key", "keypath", keypath, "err", err)
```

Which will make something like

```
2024/01/20 14:300 ERROR Unable to read ssh key keypath=/foo/bar error=file not found
```

What you've done will cause

```
2024/01/20 14:300 ERROR Unable to read ssh key: %v !BADKEY=no such file
```

See https://go.dev/blog/slog and https://pkg.go.dev/golang.org/x/exp/slog#Logger.Log for details

---

Aside: we should lint against this. CC #31